kernel/iommu/src/regs.rs

```diff
@@ -0,0 +1,212 @@
+    /// Unimplemented (may be architecturally defined)
+    _unimplemented:         [u8; 4096-0x20],   // 0x20-0xFFF
+}
+// TODO: Hardware may use more than 4kB, which means the registers may occupy
+// more than one contiguous page.
+const_assert_eq!(core::mem::size_of::<IntelIommuRegisters>(), 4096);
+
+/// Helper struct for decoding and printing capability register
+pub struct Capability(pub u64);
+
+impl Capability {
+    fn esrtps(&self)  -> bool { (self.0 >> 63) & 0x1 == 0x1 }
+    fn esirtps(&self) -> bool { (self.0 >> 62) & 0x1 == 0x1 }
+    fn fl5lp(&self)   -> bool { (self.0 >> 60) & 0x1 == 0x1 }
+    fn pi(&self)      -> bool { (self.0 >> 59) & 0x1 == 0x1 }
+    fn fl1gp(&self)   -> bool { (self.0 >> 56) & 0x1 == 0x1 }
+    fn drd(&self)     -> bool { (self.0 >> 55) & 0x1 == 0x1 }
+    fn dwd(&self)     -> bool { (self.0 >> 54) & 0x1 == 0x1 }
+    fn mamv(&self)    -> u64  { (self.0 >> 48) & 0x3f }
+    fn nfr(&self)     -> u64  { ((self.0 >> 40) & 0xff) + 1 }
+    fn psi(&self)     -> bool { (self.0 >> 39) & 0x1 == 0x1 }
+    fn sllps(&self)   -> u64  { (self.0 >> 34) & 0xf }
+    fn fro(&self)     -> u64  { (self.0 >> 24) & 0x3ff }
+    fn zlr(&self)     -> bool { (self.0 >> 22) & 0x1 == 0x1 }
+    fn mgaw(&self)    -> u64  { ((self.0 >> 16) & 0x3f) + 1 }
+    fn sagaw(&self)   -> u64  { (self.0 >> 8) & 0x1f }
+    fn cm(&self)      -> bool { (self.0 >> 7) & 0x1 == 0x1 }
+    fn phmr(&self)    -> bool { (self.0 >> 6) & 0x1 == 0x1 }
+    fn plmr(&self)    -> bool { (self.0 >> 5) & 0x1 == 0x1 }
+    fn rwbf(&self)    -> bool { (self.0 >> 4) & 0x1 == 0x1 }
+    fn afl(&self)     -> bool { (self.0 >> 3) & 0x1 == 0x1 }
+    fn nd(&self)      -> u64  { self.0 & 0x7 }
+}
+
+impl core::fmt::Display for Capability {
+    fn fmt(&self, f: &mut core::fmt::Formatter) -> core::fmt::Result {
+        // TODO: Do we really want multiple lines, since this interferes with 
+        // logging interface?
+        writeln!(f, "ND:      {}", self.nd())?;
+        writeln!(f, "AFL:     {}", self.afl())?;
+        writeln!(f, "RWBF:    {}", self.rwbf())?;
+        writeln!(f, "PLMR:    {}", self.plmr())?;
+        writeln!(f, "PHMR:    {}", self.phmr())?;
+        writeln!(f, "CM:      {}", self.cm())?;
+        writeln!(f, "SAGAW:   {}", self.sagaw())?;
+        writeln!(f, "MGAW:    {}", self.mgaw())?;
+        writeln!(f, "ZLR:     {}", self.zlr())?;
+        writeln!(f, "FRO:     {}", self.fro())?;
+        writeln!(f, "SLLPS:   {}", self.sllps())?;
+        writeln!(f, "PSI:     {}", self.psi())?;
+        writeln!(f, "NFR:     {}", self.nfr())?;
+        writeln!(f, "MAMV:    {}", self.mamv())?;
+        writeln!(f, "DWD:     {}", self.dwd())?;
+        writeln!(f, "DRD:     {}", self.drd())?;
+        writeln!(f, "FL1GP:   {}", self.fl1gp())?;
+        writeln!(f, "PI:      {}", self.pi())?;
+        writeln!(f, "FL5LP:   {}", self.fl5lp())?;
+        writeln!(f, "ESIRTPS: {}", self.esirtps())?;
+        writeln!(f, "ESRTPS:  {}", self.esrtps())?;
```

Comment:
for this, check out [`debug_struct`](https://doc.rust-lang.org/std/fmt/struct.Formatter.html#method.debug_struct)